### PR TITLE
fix(blank email): error message for blank email query parameter fixed

### LINKED ISSUE #6880: Allow blank "email" query parameter
I noticed while testing https://phabricator.services.mozilla.com/D16120 (the new about:newinstall page being developed) that they allow the self-hosted email-first form to be submit w/o requiring an email. That's a little odd, but worse is when FxA opens we show an "Invalid parameter: email" error.

screenshot 2019-01-16 at 13 17 26

STRs:

1. Open https://accounts.firefox.com/?service=sync&action=email&context=fx_desktop_v3&email=

### PR DESCRIPTION
**Pull request by farhan787**
fixes #6880

### CODE REVIEW

**Comment by farhan787** [comment]
@shane-tomlinson please review it!

File: app/scripts/lib/transform.js [hunk 1]
```diff
@@ -33,13 +51,21 @@ define(function (require, exports, module) {
     transformUsingSchema (data, schema, Errors) {
       var result = Vat.validate(data, schema);
       var error = result.error;
-      if (error instanceof ReferenceError) {
-        throw Errors.toMissingParameterError(error.key);
-      } else if (error instanceof TypeError) {
-        throw Errors.toInvalidParameterError(error.key);
-      }
+      var emptyEmail = checkEmptyEmail(data);
```
Comment (shane-tomlinson) [inline app/scripts/lib/transform.js]: We try to keep this file as generic as possible and define specifics in validation schema in the [relier models](https://github.com/mozilla/fxa-content-server/blob/c1cb23dd64011e582bb32a1395fd334987c8f9ea/app/scripts/models/reliers/relier.js#L33). I think for this case it should be sufficient to change [this line](https://github.com/mozilla/fxa-content-server/blob/c1cb23dd64011e582bb32a1395fd334987c8f9ea/app/scripts/models/reliers/relier.js#L54) to be `Vat.string().allow('')`

Comment (farhan787) [reply app/scripts/lib/transform.js]: @shane-tomlinson I was also thinking the same, it was a quick fix for this. Should I only change Vat.string().min(1) to Vat.string().allow('') ?

**Comment by farhan787** [comment]
> by not updating lib/transform, but rather only updating the validation schema in models/base/relier.js.

@shane-tomlinson but by updating the validation schema it will set error to validation error in transform.js which is again an instance of TypeError thus again executing Errors.toInvalidParameterError(error.key) function. Correct me if I am wrong.

File: app/scripts/lib/transform.js (hunk 1 above)
Comment (farhan787) [reply app/scripts/lib/transform.js]: I got it @shane-tomlinson !

**Comment by farhan787** [comment]
@shane-tomlinson I learned from my mistake, though I knew that this was specific to emails but I wanted to fix this quickly but from now I'll make sure that I work as much as generic as I can in my upcoming work. I also request you to review my pending PRs so that if there are any changes to make then I can work on them.

**Comment by shane-tomlinson** [comment]
@farhan787 - It is strange to me that the ['' unit test still passes](https://github.com/mozilla/fxa-content-server/blob/c1cb23dd64011e582bb32a1395fd334987c8f9ea/app/tests/spec/models/reliers/relier.js#L157
). `''` should be moved to be part of the [testValidQueryParam](https://github.com/mozilla/fxa-content-server/blob/c1cb23dd64011e582bb32a1395fd334987c8f9ea/app/tests/spec/models/reliers/relier.js#L161) list and I'd expect it to fail where it is.

Also, can you essentially copy [this functional test](https://github.com/mozilla/fxa-content-server/blob/c1cb23dd64011e582bb32a1395fd334987c8f9ea/tests/functional/sync_v3_email_first.js#L257)  to a new one just below that specifies an empty string for the [email](https://github.com/mozilla/fxa-content-server/blob/c1cb23dd64011e582bb32a1395fd334987c8f9ea/tests/functional/sync_v3_email_first.js#L263)?

File: app/scripts/views/index.js [hunk 2]
```diff
@@ -70,7 +70,7 @@ class IndexView extends FormView {
     if (accountFromPreviousScreen) {
       // intentionally empty
       // shows the email-first template, the prefill email was set in beforeRender
-    } else if (relierEmail) {
+    } else if (typeof(relierEmail) === 'string') {
```
Comment (shane-tomlinson) [inline app/scripts/views/index.js]: Can you remove the `()` around `relierEmail`? They aren't needed in JS.

**Comment by farhan787** [comment]
> ```js
> if (! this.get('email') && this.get('loginHint')) {
> ```

@shane-tomlinson this line is failing UI tests

**Comment by farhan787** [comment]
@shane-tomlinson I've made the changes, hope you like it!

**Comment by farhan787** [comment]
> Thank you very much for this!

@shane-tomlinson it's my pleasure :-)

**Comment by SorinaFlorean** [comment]
Verified as fixed on stage - Train 133, following the steps from #6880.

**Pull request merged**